### PR TITLE
Define Buzzer HIL and PWM Buzzer Driver

### PR DESCRIPTION
**Pull request by TeodoraMiu**
### Pull Request Overview

This pull request changes the way the buzzer driver works. Before this PR, the buzzer syscall driver assumed all buzzer systems work with PWM. This PR:
- defines a buzzer HIL
- provides a PWM buzzer driver that implements the buzzer HIL
- provides a generic buzzer syscall driver
- uses the driver for the Microbit, clue and acd boards

### Testing Strategy

This pull request was tested using a Microbit.

### TODO or Help Wanted

Feedback for the buzzer HIL is much appreciated.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: boards/microbit_v2/Cargo.toml [hunk 1]
```diff
@@ -8,7 +8,7 @@ edition = "2021"
 [dependencies]
 cortexm4 = { path = "../../arch/cortex-m4" }
 capsules = { path = "../../capsules" }
-kernel = { path = "../../kernel" }
+kernel = { path = "../../kernel", features = ["trace_syscalls"] }
```
Comment (alexandruradovici) [inline boards/microbit_v2/Cargo.toml:11]:
```suggestion
kernel = { path = "../../kernel" }
```

File: capsules/src/buzzer.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                } else {
+                    // If there is no active app or the same app is trying to use the buzzer,
+                    // we set/replace the frequency and duration.
+                    self.active_app.set(appid);
+                    self.buzzer.buzz(data1, data2).into()
+                }
+            }
+
+            2 =>
+            // Stop the current sound.
```
Comment (alexandruradovici) [inline capsules/src/buzzer.rs:111]:
```suggestion
            // Stop the current sound.
            2 =>
```

File: capsules/src/buzzer.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    ) -> CommandReturn {
+        match command_num {
+            0 =>
+            // Check whether the driver exists.
+            {
+                CommandReturn::success()
+            }
+
+            1 =>
+            // Play a sound.
```
Comment (alexandruradovici) [inline capsules/src/buzzer.rs:97]:
```suggestion
            // Play a sound.
            1 =>
```

File: capsules/src/buzzer.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    fn command(
+        &self,
+        command_num: usize,
+        data1: usize,
+        data2: usize,
+        appid: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            0 =>
+            // Check whether the driver exists.
```
Comment (alexandruradovici) [inline capsules/src/buzzer.rs:91]:
```suggestion
            // Check whether the driver exists.
            0 =>
```

File: capsules/src/buzzer.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        grant: Grant<(), UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> Buzzer<'a, B> {
+        Buzzer {
+            buzzer: buzzer,
+            apps: grant,
+            active_app: OptionalCell::empty(),
+        }
+    }
+
+    pub fn is_valid_app(&self, appid: ProcessId) -> bool {
```
Comment (alexandruradovici) [inline capsules/src/buzzer.rs:51]: Please add a comment to explain what a valid app means for this driver.

File: kernel/src/hil/buzzer.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+use crate::ErrorCode;
+
+pub trait BuzzerClient {
+    /// Called when the current sound played by the buzzer has finished
+    /// or it was stopped.
+    fn buzzer_done(&self, status: Result<(), ErrorCode>);
+}
+
+pub trait Buzzer<'a> {
```
Comment (alexandruradovici) [inline kernel/src/hil/buzzer.rs:11]: Please add a comment to explain that the main purpose of the Buzzer HIL is. (ex: playing a sound of a fixed frequency).

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: capsules/src/buzzer.rs (hunk 5 above)
Comment (TeodoraMiu) [reply capsules/src/buzzer.rs:55]: Done

File: capsules/src/buzzer.rs (hunk 4 above)
Comment (TeodoraMiu) [reply capsules/src/buzzer.rs]: Done

File: capsules/src/buzzer.rs (hunk 3 above)
Comment (TeodoraMiu) [reply capsules/src/buzzer.rs]: Done

**Comment by lschuermann** [comment]
> Before this PR, the buzzer syscall driver assumed all buzzer systems work with PWM.

Do you have an example for a buzzer which does not work with PWM? I I'm not mistaken, this PR only has an implementation for PWM-controlled buzzers, right?

**Comment by alexandruradovici** [comment]
> Do you have an example for a buzzer which does not work with PWM? I I'm not mistaken, this PR only has an implementation for PWM-controlled buzzers, right?

You might be right, even tough I can think of having for instance an external device over some bus. 

The main idea behind the HIL came from a use case where the buzzer was needed from another capsule.

**Comment by lschuermann** [comment]
I am not necessarily averse to having a buzzer HIL, but I think that there should be a compelling usecase for such peripherals which might be representable by a more general-purpose API. I figure an external chip will use some communication bus, which might be a more appropriate layer of abstraction in the kernel. Would you mind describing your usecase for controlling buzzers from within the kernel?

File: kernel/src/hil/buzzer.rs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    /// Stop the sound currenty playing.
+    /// After the buzzer is successfully stopped, the `buzzer_done()`
+    /// callback is called.
+    ///
+    /// Return values:
+    ///
+    /// - `Ok(())`: The attempt at stopping the buzzer was successful.
+    /// - `FAIL`: Cannot stop the buzzer.
+    fn stop(&self) -> Result<(), ErrorCode>;
```
Comment (bradjc) [inline kernel/src/hil/buzzer.rs:34]: If the buzzer is not currently buzzing, what happens if `stop()` is called?

**Review by bradjc — CHANGES_REQUESTED** [review]
Why did you remove the support for multiple apps and queuing?

**Comment by TeodoraMiu** [comment]
> Why did you remove the support for multiple apps and queuing?

I figured that an app that wants to notify the user using the buzzer wants to do it immediately, not wait after another notification. Therefore the buzzer should play the sound instantly, otherwise it doesn't really serve its purpose.

File: kernel/src/hil/buzzer.rs (hunk 7 above)
Comment (TeodoraMiu) [reply kernel/src/hil/buzzer.rs]: I changed it so that the capsule returns ErrorCode::OFF if the stop command is called while the buzzer isn't buzzing.

**Review by alexandruradovici — APPROVED** [review]

**Comment by phil-levis** [comment]
@bradjc Have your concerns been addressed?

**Comment by bradjc** [comment]
> @bradjc Have your concerns been addressed?

No, not really. We currently have a virtualized buzzer that supports multiple apps, and this replaces it with one that makes the userspace app responsible for doing retries. Perhaps we need a new "buzz immediate" syscall or syscall driver that boards that only want immediate buzz (or if the platform can't provide an immediate buzz it returns an error) can use. But I don't see why we would get rid of the current support for queuing in the process.

I also don't see the need to change the name from `buzzer_driver.rs` to `buzzer.rs`. `buzzer_driver` is more clear this is a userspace driver.

**Comment by TeodoraMiu** [comment]
I believe this is now good to go. I changed the driver so that there is a separate syscall for the buzzer to immediately buzz, but I also kept a syscall for queuing buzz commands. I also returned to the old buzzer_driver.rs name.

**Comment by phil-levis** [comment]
> I believe this is now good to go. I changed the driver so that there is a separate syscall for the buzzer to immediately buzz, but I also kept a syscall for queuing buzz commands. I also returned to the old buzzer_driver.rs name.

@bradjc?

**Comment by phil-levis** [comment]
@TeodoraMiu Looks like there are some simple conflicts to resolve.

File: kernel/src/hil/buzzer.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    /// Once the buzzer finishes buzzing, the `buzzer_done()` callback
+    /// is called.
+    /// If it is called while the buzzer is playing, the buzzer command will be
+    /// overriden with the new frequency and duration values.
+    ///
+    /// Return values:
+    ///
+    /// - `Ok(())`: The attempt at starting the buzzer was successful.
+    /// - `FAIL`: Cannot start the buzzer.
+    fn buzz(&self, frequency_hz: usize, duration_ms: usize) -> Result<(), ErrorCode>;
```
Comment (ppannuto) [inline kernel/src/hil/buzzer.rs:24]: Should this have an `EVALUE` for impossible frequencies?

**Review by bradjc — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Comment by phil-levis** [comment]
Great!

**Pull request merged**